packages/framer-motion/src/render/VisualElement.ts

```diff
@@ -0,0 +1,810 @@
+     * value, we'll create one if none exists.
+     */
+    getValue(key: string, defaultValue?: string | number) {
+        if (this.props.values && this.props.values[key]) {
+            return this.props.values[key]
+        }
+
+        let value = this.values.get(key)
+
+        if (value === undefined && defaultValue !== undefined) {
+            value = motionValue(defaultValue)
+            this.addValue(key, value)
+        }
+
+        return value as MotionValue
+    }
+
+    /**
+     * If we're trying to animate to a previously unencountered value,
+     * we need to check for it in our state and as a last resort read it
+     * directly from the instance (which might have performance implications).
+     */
+    readValue(key: string) {
+        return this.latestValues[key] !== undefined || !this.current
+            ? this.latestValues[key]
+            : this.readValueFromInstance(this.current, key, this.options)
+    }
+
+    /**
+     * Set the base target to later animate back to. This is currently
+     * only hydrated on creation and when we first read a value.
+     */
+    setBaseTarget(key: string, value: string | number) {
+        this.baseTarget[key] = value
+    }
+
+    /**
+     * Find the base target for a value thats been removed from all animation
+     * props.
+     */
+    getBaseTarget(key: string) {
+        const { initial } = this.props
+        const valueFromInitial =
+            typeof initial === "string" || typeof initial === "object"
+                ? resolveVariantFromProps(this.props, initial as any)?.[key]
+                : undefined
+
+        /**
+         * If this value still exists in the current initial variant, read that.
+         */
+        if (initial && valueFromInitial !== undefined) {
+            return valueFromInitial
+        }
+
+        /**
+         * Alternatively, if this VisualElement config has defined a getBaseTarget
+         * so we can read the value from an alternative source, try that.
+         */
+        const target = this.getBaseTargetFromProps(this.props, key)
+        if (target !== undefined && !isMotionValue(target)) return target
+
+        /**
+         * If the value was initially defined on initial, but it doesn't any more,
+         * return undefined. Otherwise return the value as initially read from the DOM.
+         */
+        return this.initialValues[key] !== undefined &&
+            valueFromInitial === undefined
+            ? undefined
+            : this.baseTarget[key]
+    }
+
+    on<EventName extends keyof VisualElementEventCallbacks>(
+        eventName: EventName,
+        callback: VisualElementEventCallbacks[EventName]
+    ) {
+        if (!this.events[eventName]) {
+            this.events[eventName] = new SubscriptionManager()
+        }
+
+        return this.events[eventName].add(callback)
+    }
+
+    notify<EventName extends keyof VisualElementEventCallbacks>(
```

Comment:
Does it need to be `extends`?
Don't we always want it to be one of the `VisualElementEventCallbacks`?

Reply:
Yeah I'm not sure how else to type it though. By doing it like this I can send `EventName` through to `callback: VisualElementEventCallbacks[EventName]` and get the correct type for that.